server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java

```diff
@@ -0,0 +1,415 @@
+        if (intervalType.equals(IntervalTypeEnum.FIXED) || tryIntervalAsFixedUnit() != null) {
+            return dateHistogramInterval;
+        }
+        return null;
+    }
+
+    /**
+     * Sets the interval of the DateHistogram using fixed units (`1ms`, `1s`, `10m`, `4h`, etc).  These are
+     * not calendar aware and are simply multiples of fixed, SI units.
+     *
+     * This is mutually exclusive with {@link DateIntervalWrapper#calendarInterval(DateHistogramInterval)}
+     *
+     * @param interval The fixed interval to use
+     */
+    public void fixedInterval(DateHistogramInterval interval) {
+        if (interval == null || Strings.isNullOrEmpty(interval.toString())) {
+            throw new IllegalArgumentException("[interval] must not be null: [date_histogram]");
+        }
+        setIntervalType(IntervalTypeEnum.FIXED);
+        // Parse to make sure it is a valid fixed too
+        TimeValue.parseTimeValue(interval.toString(), DateHistogramAggregationBuilder.NAME + ".fixedInterval");
+        this.dateHistogramInterval = interval;
+    }
+
+    /** Return the interval as a date time unit if applicable, regardless of how it was configured. If this returns
+     *  {@code null} then it means that the interval is expressed as a fixed
+     *  {@link TimeValue} and may be accessed via {@link #tryIntervalAsFixedUnit()}. */
+    DateTimeUnit tryIntervalAsCalendarUnit() {
+        if (intervalType.equals(IntervalTypeEnum.CALENDAR) || intervalType.equals(IntervalTypeEnum.LEGACY_DATE_HISTO)) {
+            return DateHistogramAggregationBuilder.DATE_FIELD_UNITS.get(dateHistogramInterval.toString());
+        }
+        return null;
+    }
+
+    /**
+     * Get the interval as a {@link TimeValue}, regardless of how it was configured. Returns null if
+     * the interval cannot be parsed as a fixed time.
+     */
+    TimeValue tryIntervalAsFixedUnit() {
+        if (dateHistogramInterval == null || Strings.isNullOrEmpty(dateHistogramInterval.toString())) {
+            return null;
+        }
+        try {
+            return TimeValue.parseTimeValue(dateHistogramInterval.toString(), null, getClass().getSimpleName() + ".interval");
+        } catch (IllegalArgumentException e) {
+            return null;
+        }
+    }
+
+    public Rounding createRounding(ZoneId timeZone) {
+        Rounding.Builder tzRoundingBuilder;
+        if (isEmpty()) {
+            throw new IllegalArgumentException("Invalid interval specified, must be non-null and non-empty");
+        }
+        DateIntervalWrapper.IntervalTypeEnum intervalType = getIntervalType();
+        if (intervalType.equals(DateIntervalWrapper.IntervalTypeEnum.FIXED)) {
+            tzRoundingBuilder = Rounding.builder(tryIntervalAsFixedUnit());
+        } else if (intervalType.equals(DateIntervalWrapper.IntervalTypeEnum.CALENDAR)) {
+            tzRoundingBuilder = Rounding.builder(tryIntervalAsCalendarUnit());
+        } else {
```

Comment:
The `else` section is basically both `LEGACY` branches, using the old parsing logic.  That's why it retains the "try as calendar first" logic, and then falls back to a fixed interval if it can't parse as a calendar.  

The old logic is setup that way since `LEGACY_DATE_HISTO` can be both calendar and fixed. :(